android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java

```diff
@@ -46,37 +53,100 @@ public void onCreate(Bundle savedInstanceState) {
     getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
     // Load the preferences from an XML resource
     addPreferencesFromResource(R.xml.settings);
+    updatePwsList();
+  }
+
+  @Override
+  public void onResume() {
+    super.onResume();
+    getActivity().getActionBar().setTitle(R.string.title_settings);
+  }
+
+  @Override
+  public void onPrepareOptionsMenu(Menu menu) {
+    super.onPrepareOptionsMenu(menu);
+    menu.findItem(R.id.action_config).setVisible(false);
+    menu.findItem(R.id.action_about).setVisible(false);
+    menu.findItem(R.id.action_settings).setVisible(false);
+  }
+
+  @Override
+  public boolean onPreferenceTreeClick(PreferenceScreen preferenceScreen, Preference preference){
+    super.onPreferenceTreeClick(preferenceScreen, preference);
+
+    if (preference instanceof PreferenceScreen){
+      setupHomeButton(((PreferenceScreen) preference));
+    }
+
+    return false;
+  }
+
+  private void setupHomeButton(PreferenceScreen preference){
+    final Dialog dialog = preference.getDialog();
+    dialog.getActionBar().setDisplayHomeAsUpEnabled(true);
+
+    View homeBtn = dialog.findViewById(android.R.id.home);
+    if (homeBtn != null) {
+      OnClickListener dismissDialogClickListener = new OnClickListener() {
+          @Override
+          public void onClick(View v) {
+            updatePwsPreference();
+            updatePwsList();
+            dialog.dismiss();
+          }
+      };
+      ((ViewGroup) homeBtn.getParent().getParent()).setOnClickListener(dismissDialogClickListener);
```

Comment:
This is pretty brittle as it relies on details of the ActionBar layout that we don't have control over.  What happens if we don't setDisplayHomeAsUpEnabled(true)?  Can you still navigate back using the system Back button?